spec/services/arcgis_api/token_keeper_spec.rb

```diff
@@ -0,0 +1,219 @@
+  # end
+
+  before(:each) do
+    allow(Rails).to receive(:cache).and_return(cache_store)
+    subject.remove_token
+  end
+
+  shared_examples 'acquire token test' do
+    context 'token not expired and not in prefetch timeframe' do
+      it 'get same token at second call' do
+        expected = 'ABCDEFG'
+        expected_sec = 'GFEDCBA'
+
+        stub_request(:post, %r{/generateToken}).to_return(
+          { status: 200,
+            body: {
+              token: expected,
+              expires: (Time.zone.now.to_f + 5) * 1000,
+              ssl: true,
+            }.to_json,
+            headers: { content_type: 'application/json;charset=UTF-8' } },
+          { status: 200,
+            body: {
+              token: expected_sec,
+              expires: (Time.zone.now.to_f + 3600) * 1000,
+              ssl: true,
+            }.to_json,
+            headers: { content_type: 'application/json;charset=UTF-8' } },
+        )
+
+        expect(Rails.cache).to receive(:read).with(kind_of(String)).
+          and_call_original
+        token = subject.token
+        expect(token).to eq(expected)
+        sleep(1)
+        expect(Rails.cache).to receive(:read).with(kind_of(String)).
+          and_call_original
+        token = subject.token
+        expect(token).to eq(expected)
+      end
+    end
+
+    context 'token not expired and but in prefetch timeframe' do
+      let(:expected) { 'ABCDEFG' }
+      let(:expected_sec) { 'GFEDCBA' }
+      before(:each) do
+        stub_request(:post, %r{/generateToken}).to_return(
+          { status: 200,
+            body: {
+              token: expected,
+              expires: (Time.zone.now.to_f + 15) * 1000,
+              ssl: true,
+            }.to_json,
+            headers: { content_type: 'application/json;charset=UTF-8' } },
+          { status: 200,
+            body: {
+              token: expected_sec,
+              expires: (Time.zone.now.to_f + 3600) * 1000,
+              ssl: true,
+            }.to_json,
+            headers: { content_type: 'application/json;charset=UTF-8' } },
+        )
+      end
+      let(:prefetch_ttl) do
+        5
+      end
+      context 'get token at different timing' do
+        it 'get same token between sliding_expires_at passed and sliding_expires_at+prefetch_ttl' do
+          expect(Rails.cache).to receive(:read).with(kind_of(String)).
+            and_call_original
+          token = subject.token
+          expect(token).to eq(expected)
+          sleep(1)
+          expect(Rails.cache).to receive(:read).with(kind_of(String)).
+            and_call_original
+          token = subject.token
+          expect(token).to eq(expected)
+        end
+        it 'regenerates token when passed sliding_expires_at+prefetch_ttl' do
+          expect(Rails.cache).to receive(:read).with(kind_of(String)).
+            and_call_original
+          token = subject.token
+          expect(token).to eq(expected)
+          sleep(11)
```

Comment:
We should avoid calling `sleep` in tests because it can be quite slow.

Reply:
Or at least stub `Kernel#sleep` so it doesn't take wall clock time

We can also use gems like [`#with_retries`](https://github.com/ooyala/retries) which have good test support and don't take up wall clock time either

Reply:
checked the existing tests, there are helper to freeze time and travel to new time as needed. will use that.

Reply:
No :sleeping: anymore